force-app/main/default/classes/ExchangeRateProvider.cls

```
@@ -0,0 +1,409 @@
+            // Get sequence number
+            String seqStr = root.getAttribute('poradi', null);
+            if (String.isNotBlank(seqStr)) {
+                cnbData.sequenceNumber = Integer.valueOf(seqStr);
+            }
+            
+            // Parse exchange rates
+            List<Dom.XmlNode> rateNodes = root.getChildElements();
+            for (Dom.XmlNode rateNode : rateNodes) {
+                if (rateNode.getName() == 'radek') {
+                    CNBExchangeRate rate = parseExchangeRateNode(rateNode);
+                    if (rate != null) {
+                        cnbData.rates.add(rate);
+                    }
+                }
+            }
+            
+            return cnbData;
+            
+        } catch (Exception e) {
+            throw new ExchangeRateException(ERROR_XML_PARSING + ': ' + e.getMessage());
+        }
+    }
+    
+    /**
+     * @description Parses individual exchange rate node
+     * @param rateNode XML node containing exchange rate data
+     * @return CNBExchangeRate object
+     */
+    private static CNBExchangeRate parseExchangeRateNode(Dom.XmlNode rateNode) {
+        try {
+            CNBExchangeRate rate = new CNBExchangeRate();
+            
+            rate.country = rateNode.getAttribute('zeme', '');
+            rate.currencyName = rateNode.getAttribute('mena', '');
+            rate.amount = Integer.valueOf(rateNode.getAttribute('mnozstvi', '1'));
+            rate.currencyCode = rateNode.getAttribute('kod', '');
+            rate.exchangeRate = Decimal.valueOf(rateNode.getAttribute('kurz', '0'));
+            
+            // Validate required fields
+            if (String.isBlank(rate.currencyCode) || rate.exchangeRate == 0) {
+                return null;
+            }
+            
+            return rate;
+            
+        } catch (Exception e) {
+            System.debug(LoggingLevel.WARN, 'Failed to parse exchange rate node: ' + e.getMessage());
+            return null;
```

🛑 **[Robustness] Refactor to Throw Exceptions on Failure Instead of Returning Null**

Silent data loss issue discussed in the parent parseCNBXMLResponse method. By returning null when parsing fails (either in the catch block or during validation), it effectively "swallows" critical errors. Instead, it should throw a custom exception if the data is invalid or a parsing error occurs or concatenate the exception to the main one.